pkg/registry/core/pod/strategy.go

```diff
@@ -114,6 +114,9 @@ func (podStrategy) WarningsOnCreate(ctx context.Context, obj runtime.Object) []s
 
 // Canonicalize normalizes the object after validation.
 func (podStrategy) Canonicalize(obj runtime.Object) {
+	// This has to happen here, rather than PrepareForCreate() because
+	// generateName has not yet been evaluated there.
+	setDefaultHostname(obj.(*api.Pod))
```

Comment:
what is the fallout? 
I mean, we want to fix deployments but what other things use this codepath and will be affected?

Reply:
This should be OK.  If you set this field manually, we take your value.  Otherwise this is the same as `Kubelet.GeneratePodHostNameAndDomain` and statefulset `initIdentity`.